### PR TITLE
Adds new CLI features and IOPointer unflag all option

### LINKED ISSUE #157: CLI command list
Currently we have:

- `recent`
- `history`
- `trace`

We would also like to include:

- `inspect COMPONENT_RUN_ID`
- `ls tags`
- `ls components`
- `ls components --tag=TAG`
- `ls components --owner=NAME`

We can use `get_components_for_owner` and `get_components_for_tag` to make the `ls` commands.

### LINKED ISSUE #203: [Feature] Ability to "unflag all"
It is probably tedious to go and unflag all flagged outputs. Add a command to unflag all.

### PR DESCRIPTION
**Pull request by aditim1359**
Addressed the cli features: addd ml trace tags, components and unflag all

Closes #203 
Closes #157

### CODE REVIEW

File: mltrace/cli/cli.py [hunk 1]
```diff
@@ -223,6 +227,22 @@ def recent(limit: int, address: str = ""):
...
+@click.argument("component_run_id")
+def inspect(component_run_id, address: str = ""):
+    """
+    CLI to inspect a specific component run id.
+    """
+    # Set address
+    if address and len(address) > 0:
+        set_address(address)
+    # Get the recent ids
+    component_run_ids = get_recent_run_ids()
```
Comment (shreyashankar) [inline mltrace/cli/cli.py:241]: Your logic here only shows info for the component run ID if it was in the most recent run IDs. We want to be able to show info regardless. You don't need the if statement -- you can just directly call `show_info_card`

File: mltrace/cli/cli.py [hunk 2]
```diff
@@ -274,16 +294,27 @@ def flag(output_id: str, address: str = ""):
 
 
 @mltrace.command("unflag")
-@click.argument("output_id")
+@click.option("--output_id", help="Output ID to unflag")
+@click.option("--all", is_flag=True)
```
Comment (shreyashankar) [inline mltrace/cli/cli.py:298]: What is `is_flag = True`? You can do `@click.option("--all", type=bool, default-False)`

File: mltrace/cli/cli.py [hunk 3]
```diff
@@ -274,16 +294,27 @@ def flag(output_id: str, address: str = ""):
...
 @click.option("--address", help="Database server address")
-def unflag(output_id: str, address: str = ""):
+def unflag(output_id: str = "", all: bool = False, address: str = ""):
```
Comment (shreyashankar) [inline mltrace/cli/cli.py:300]: What if they set both `all` and an `output_id`? Check for this in the beginning and raise an error.

File: mltrace/cli/cli.py [hunk 4]
```diff
@@ -307,3 +338,67 @@ def review(limit: int = 5, address: str = ""):
...
+    if address and len(address) > 0:
+        set_address(address)
+
+    # Make return result
+    components = []
+    if owner:
+        components = get_components_with_owner(owner)
+    elif tag:
+        components = get_components_with_tag(tag)
+    else:
```
Comment (shreyashankar) [inline mltrace/cli/cli.py:360]: What is this else case for? I'm not sure it's necessary

File: mltrace/cli/cli.py [hunk 5]
```diff
@@ -307,3 +338,67 @@ def review(limit: int = 5, address: str = ""):
...
+def tags(address: str = ""):
+    """
+    Command to set the flag property of an output_id to true.
+    """
+    # Set address
+    if address and len(address) > 0:
+        set_address(address)
+
+    # Get all ids
+    component_run_ids = get_all_run_ids()
```
Comment (shreyashankar) [inline mltrace/cli/cli.py:394]: Not an efficient way to do this. Querying the db directly for tags will be faster. To do this:

1) Make a method in `store.py` to query the tags

```
def get_all_tags(self) -> typing.List[Tag]:
  return self.session.query(Tag).all()
```

2. Create wrapper in `client.py` and whitelist the wrapper in `mltrace/__init__.py`

```
def get_all_tags()  -> typing.List[Tag]:
  store = Store(_db_uri)
  res = store.get_all_tags()
  # Convert res to client-side Tag objects & return
```

3. Use `client.py` function in `cli.py`

File: mltrace/client.py [hunk 6]
```diff
@@ -443,6 +443,12 @@ def get_recent_run_ids(limit: int = 5, last_run_id=None):
     return store.get_recent_run_ids(limit, last_run_id)
 
 
+def get_all_run_ids(last_run_id=None):
```
Comment (shreyashankar) [inline mltrace/client.py:446]: This will only get the last 50 run ids by default. Also no need for this function

**Review by shreyashankar — CHANGES_REQUESTED** [review]

File: mltrace/cli/cli.py [hunk 7]
```diff
@@ -274,16 +294,27 @@ def flag(output_id: str, address: str = ""):
...
-    Command to set the flag property of an output_id to false.
+    Command to set flag property of an output_id or all output_ids to false.
     """
     # Set address
     if address and len(address) > 0:
         set_address(address)
-    unflag_output_id(output_id)
+
+    if all:
+        outputs, component_counts = review_flagged_outputs()
```
Comment (shreyashankar) [inline mltrace/cli/cli.py:309]: Write a `store.py` and `client.py` function to directly get all the flagged iops and delete them:

```
flagged_iops = (
            self.session.query(IOPointer)
            .filter(IOPointer.flag.is_(True))
            .all()
        )

for iop in flagged_iops:
  iop.clear_flag()

self.session.commit()
```

File: mltrace/cli/cli.py (hunk 1 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:241]: done.

File: mltrace/cli/cli.py (hunk 2 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:298]: oh ok, I was researching how to do boolean flags with click and this was the option I found that worked best. I didn't know you can set type.

File: mltrace/cli/cli.py (hunk 3 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:300]: done.

File: mltrace/cli/cli.py (hunk 4 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:360]: It is for the general case of if they did not set an owner or a tag, list all components.

File: mltrace/cli/cli.py (hunk 5 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:394]: done

File: mltrace/client.py (hunk 6 above)
Comment (aditim1359) [reply mltrace/client.py:446]: removed this and implemented the cli feature differently by directly requiring the db

File: mltrace/cli/cli.py (hunk 7 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:309]: done

File: mltrace/cli/cli.py (hunk 2 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:298]: Actually! I remember why I did this. this way, if the flag "--all" is passed on the cl, then it is set to True other wise it is False. ie don't have to type --all=True. Easier and more intuitive use

**Comment by aditim1359** [comment]
The reason the tests are failing are because of it testing te functionality of the get_components_with_owner function which I condensed into get_components. Are there any other things off the top of your head that need the get_components_with_owner and with_tag separate functions? if so, I can update them or we can discuss. Otherwise, I will update those tests to just test for get_components.

**Comment by shreyashankar** [comment]
Your solution sounds good to me! I don't think there are any dependencies on the old functions

File: mltrace/cli/cli.py [hunk 8]
```diff
@@ -307,3 +341,49 @@ def review(limit: int = 5, address: str = ""):
...
+                                   "flags passed.")
+
+    # Display components, one per line
+    for comp in result:
+        # print("accessing: ", 1)
+        click.echo(f"Name: {comp.name}")
+        click.echo(f"└─Description: {comp.description}")
+        click.echo(f"└─Owner: {comp.owner}")
+        click.echo(f"└─Tags: {comp.tags}")
+        # print("fin")
```
Comment (shreyashankar) [inline mltrace/cli/cli.py:371]: Delete line

File: mltrace/__init__.py [hunk 9]
```diff
@@ -31,15 +32,14 @@
...
-    "get_components_with_owner",
     "tag_component",
-    "get_components_with_tag",
     "log_component_run",
     "create_random_ids",
     "get_component_information",
     "get_component_run_information",
     "web_trace",
     "get_recent_run_ids",
+    "get_all_run_ids",
```
Comment (shreyashankar) [inline mltrace/__init__.py:42]: Delete line

File: mltrace/db/store.py [hunk 10]
```diff
@@ -1,582 +1,605 @@
...
+
+        trace_nodes_counts = sorted(
+            trace_nodes_counts.items(),
+            key=lambda item: (-item[1], -item[0].id),
+        )
+
+        # Return a list of the ComponentRuns in the order
+        return flagged_output_ids, trace_nodes_counts
+
+    def get_all_tags(self) -> typing.List[Tag]:
```
Comment (shreyashankar) [inline mltrace/db/store.py:604]: Why is this empty?

File: mltrace/client.py [hunk 11]
```diff
@@ -443,13 +433,30 @@ def get_recent_run_ids(limit: int = 5, last_run_id=None):
...
     """Returns IO Pointer metadata."""
     store = Store(_db_uri)
     iop = store.get_io_pointer(io_pointer_id, create)
     return IOPointer.from_dictionary(iop.__dict__)
 
 
+def get_all_tags() -> typing.List[str]:
+    store = Store(_db_uri)
+    res = store.get_all_tags()
+    tags = set()
```
Comment (shreyashankar) [inline mltrace/client.py:453]: Tags are unique so you can keep it a list. See `db/models.py` for the schema

`tags = [t.name for t in res]`

**Review by shreyashankar — CHANGES_REQUESTED** [review]
Super close! Just a couple of nits & one method stub to fill out

File: mltrace/cli/cli.py (hunk 2 above)
Comment (shreyashankar) [reply mltrace/cli/cli.py]: awesome

File: mltrace/cli/cli.py (hunk 8 above)
Comment (aditim1359) [reply mltrace/cli/cli.py:371]: done

File: mltrace/client.py (hunk 11 above)
Comment (aditim1359) [reply mltrace/client.py:453]: Oh nice ok.

File: mltrace/db/store.py (hunk 10 above)
Comment (aditim1359) [reply mltrace/db/store.py]: woops.

File: mltrace/__init__.py (hunk 9 above)
Comment (aditim1359) [reply mltrace/__init__.py:42]: done

File: mltrace/client.py [hunk 12]
```diff
@@ -443,13 +433,27 @@ def get_recent_run_ids(limit: int = 5, last_run_id=None):
     return store.get_recent_run_ids(limit, last_run_id)
 
 
+def get_all_run_ids():
```
Comment (shreyashankar) [inline mltrace/client.py:436]: This function doesn't need to exist, also `store.py` doesn't have the method `.get_all_run_ids`

**Review by shreyashankar — CHANGES_REQUESTED** [review]
So close!!

File: mltrace/client.py (hunk 12 above)
Comment (aditim1359) [reply mltrace/client.py:436]: ah right.

**Review by shreyashankar — APPROVED** [review]
LGTM! Feel free to merge :)

**Pull request merged**